wp-includes/class.wp-service-workers.php

```diff
@@ -0,0 +1,153 @@
+<?php
+/**
+ * Dependencies API: WP_Service_Workers class
+ *
+ * @since ?
+ *
+ * @package PWA
+ */
+
+/**
+ * Class used to register service workers.
+ *
+ * @since ?
+ *
+ * @see WP_Dependencies
+ * @todo Does it make sense to extend WP_Scripts, maybe WP_Dependencies is enough?
+ */
+class WP_Service_Workers extends WP_Scripts {
+
+	/**
+	 * Param for service workers.
+	 *
+	 * @var string
+	 */
+	public $query_var = 'wp_service_worker';
+
+	/**
+	 * Array of scopes.
+	 *
+	 * @var array
+	 */
+	public $scopes = array();
+
+	/**
+	 * Output for service worker scope script.
+	 *
+	 * @var string
+	 */
+	public $output = '';
+
+	/**
+	 * WP_Service_Workers constructor.
+	 */
+	public function __construct() {
+		parent::__construct();
+		global $wp_filesystem;
+
+		if ( ! class_exists( 'WP_Filesystem' ) ) {
+			require_once ABSPATH . '/wp-admin/includes/file.php';
+		}
+
+		if ( null === $wp_filesystem ) {
+			WP_Filesystem();
+		}
+	}
+
+	/**
+	 * Initialize the class.
+	 */
+	public function init() {
+		/**
+		 * Fires when the WP_Service_Workers instance is initialized.
+		 *
+		 * @param WP_Service_Workers $this WP_Service_Workers instance (passed by reference).
+		 */
+		do_action_ref_array( 'wp_default_service_workers', array( &$this ) );
+	}
+
+	/**
+	 * Register service worker.
+	 *
+	 * Registers service worker if no item of that name already exists.
+	 *
+	 * @param string $handle Name of the item. Should be unique.
+	 * @param string $path   Path of the item relative to the WordPress root directory.
+	 * @param array  $deps   Optional. An array of registered item handles this item depends on. Default empty array.
+	 * @param bool   $ver    Always false for service worker.
+	 * @param mixed  $scope  Optional. Scope of the service worker. Default relative path.
+	 * @return bool Whether the item has been registered. True on success, false on failure.
+	 */
+	public function add( $handle, $path, $deps = array(), $ver = false, $scope = null ) {
```

Comment:
Can we just remove the `$ver` entirely?

Reply:
We could if we would create a function with different name, otherwise it wouldn't be compatible with the parent method (`WP_Dependencies::add`). Do you think it would be better to create a separate method? Thoughts?

Reply:
Oh yeah. PHP will issue a notice about the function signatures not matching, correct? Alas the problem with positional parameters.

We could introduce a `register` method instead which would be the only method used instead of `add` or `enqueue`. I don't think it makes sense to ever add a service worker script and then not have it be enqueued ever, right? Setting the `scope` is really the way that you are enqueueing it.